File: MyClassLibrary/Classes/class_Functions.cs
using DevExpress.XtraEditors;
using Microsoft.VisualBasic;
using System;
using System.Linq;
using System.Text.RegularExpressions;
using System.Windows.Forms;

namespace MyClassLibrary.Classes
{
	public static class class_Functions
	{
		public enum PasswordScore
		{
			Blank,
			VeryWeak,
			Weak,
			Medium,
			Strong,
			VeryStrong
		}

		public static PasswordScore CheckPasswordStrength(string password)
		{
			int N = 0;
			int L = password.Length;
			if (L == 0)
				return PasswordScore.Blank;
			if (Regex.IsMatch(password, @"[\d]", RegexOptions.ECMAScript))
				N += 10;
			if (Regex.IsMatch(password, "[a-z]", RegexOptions.ECMAScript))
				N += 26;
			if (Regex.IsMatch(password, "[A-Z]", RegexOptions.ECMAScript))
				N += 26;
			if (Regex.IsMatch(password, @"[~`!@#$%\^\&\*\(\)\-_\+=\[\{\]\}\|\\;:'\""<\,>\.\?\/£]", RegexOptions.ECMAScript) && password.Length > 8)
				N += 33;
			int H;
			try
			{
				H = Convert.ToInt32(L * Math.Round(Math.Log(N) / Math.Log(2d)));
			}
			catch (Exception ex)
			{
				H = 0;
				class_Procedures.Show_Error(ex);
			}

			if (H <= 32)
				return PasswordScore.VeryWeak;
			if (H <= 48)
				return PasswordScore.Weak;
			if (H <= 64)
				return PasswordScore.Medium;
			if (H <= 80)
				return PasswordScore.Strong;
			return PasswordScore.VeryStrong;
		}

		public static string Product_Name_Version()
		{
			return $"{Application.ProductName} {Application.ProductVersion}™ {DateTime.Today.Year}";
		}

		public static string AutoID(string sPreCode)
		{
			return $"{sPreCode}{DateTime.Now:yyyyMMddHHmmssfff}";
		}

		public static DateTime FirstDayOfMonth(DateTime value)
		{
			return new DateTime(value.Year, value.Month, 1);
		}

		public static int DaysInMonth(this DateTime value)
		{
			return DateTime.DaysInMonth(value.Year, value.Month);
		}

		public static DateTime LastDayOfMonth(this DateTime value)
		{
			return new DateTime(value.Year, value.Month, value.DaysInMonth());
		}

		public static int GetMonthDifference(DateTime startDate, DateTime endDate)
		{
			return Math.Abs(12 * (startDate.Year - endDate.Year) + startDate.Month - endDate.Month);
		}

		public static DateTime Next_Working_Date(DateTime dt, bool bShiftPressed = false)
		{
			int i;

			i = bShiftPressed ? -1 : 1;

			dt = dt.AddDays(i);
			if (dt.DayOfWeek.Equals(DayOfWeek.Sunday))
			{
				dt = dt.AddDays(i);
			}

			return dt;
		}

		public static int WorkingDays(DateTime dtDateFrom, DateTime dtDateTo)
		{
			int _elapsedDays = 0;
			var _weekendDays = new[] { DayOfWeek.Sunday };
			for (int i = 0, loopTo = (dtDateTo - dtDateFrom).Days; i <= loopTo; i++)
			{
				if (!_weekendDays.Contains(dtDateFrom.AddDays(i).DayOfWeek))
					_elapsedDays += 1;
			}

			return _elapsedDays;
		}

		public static DateTime WorkingDays(DateTime dtDateFrom, int iDAys)
		{
			var _elapsedDate = dtDateFrom;
			var _weekendDays = new[] { DayOfWeek.Saturday, DayOfWeek.Sunday };
			if (iDAys > 0)
			{
				for (int i = 0, loopTo = iDAys - 1; i <= loopTo; i++)
				{
					_elapsedDate = _elapsedDate.AddDays(1d);
					if (_weekendDays.Contains(_elapsedDate.DayOfWeek))
					{
						i -= 1;
					}
				}
			}

			return _elapsedDate;
		}

		public static string Generate_ID(string sCode, int iZeros)
		{
			return Convert.ToInt32(sCode.Remove(0, sCode.Length - iZeros)).ToString();
		}

		public static string Generate_ID(string sPreCode, int iZeros, int iCode)
		{
			string sTmpCode, sCode;
			int iLength = iCode.ToString().Length;
			if (iLength > iZeros)
				iZeros = iLength;
			string sZeros = new string('0', iZeros);
			sTmpCode = $"{sPreCode}{sZeros}";
			sCode = sTmpCode.Remove(sTmpCode.Length - iLength);
			return sCode.Insert(sTmpCode.Length - iLength, iCode.ToString());
		}

		public static string Code_Increment(string sCode, int iCount)
		{
			if (sCode.Equals(string.Empty))
				return string.Empty;
			int iValueNew = Convert.ToInt32(sCode.Remove(0, sCode.Length - iCount)) + 1;
			return $"{sCode.Remove(sCode.Length - iValueNew.ToString().Length)}{iValueNew}";
		}

		public static double calculate_VAT(double dAmount)
		{
			return Math.Round(dAmount / 1.12 * 0.12, 2);
		}

		public static double calculate_no_VAT(double dAmount)
		{
			return Math.Round(dAmount - dAmount / 1.12 * 0.12, 2);
		}

		public static decimal Percent(decimal dAmount, decimal dRate)
		{
			//return dAmount * dRate / 100 - dAmount;
			//return (dAmount * dRate) - dAmount;
			return (dAmount * dRate);
		}

		public static string maskString(string s, int iUnMaskLeft, int iUnMaskRight)
		{
			string sReturn = null;
			try
			{
				string firstPart = s.Substring(0, iUnMaskLeft);
				int len = s.Length;
				string lastPart = s.Substring(len - iUnMaskRight, iUnMaskRight);
				int middlePartLenght = len - (firstPart.Length + lastPart.Length);
				string middlePart = new string('•', middlePartLenght);
				sReturn = $"{firstPart}{middlePart}{lastPart}";
			}
			catch (Exception ex)
			{
				class_Procedures.Show_Error(ex);
			}

			return sReturn;
		}

		public static string AmountToWords(Decimal dNumber, bool bCurrency = true)
		{
			if (dNumber <= 0) return null;

			string sNumberSet = Strings.Trim(dNumber.ToString());
			var sAmount = sNumberSet.Split(new char[] { '.' });
			string sPesos = sAmount[0];
			string sCents = null;
			try
			{
				sCents = Strings.Left(sAmount[1], 2);
			}
			catch (Exception ex)
			{
				class_Procedures.Show_Error(ex, "Looks like your Amount has no point value.", string.Empty, "Amount in Words"); //Please make sure that the [Amount in Words] are specified
			}

			try
			{
				if (Convert.ToInt32(sCents) > 0)
				{
					if (bCurrency)
						return $"{NumberToText(Convert.ToInt32(sPesos)).Trim()} Pesos and {NumberToText(Convert.ToInt32(sCents))} Cents Only";
					else
						return $"{NumberToText(Convert.ToInt32(sPesos)).Trim()} point {NumberToText(Convert.ToInt32(sCents))}";
				}
				else
				{
					if (bCurrency)
						return $"{NumberToText(Convert.ToInt32(sPesos)).Trim()} Pesos Only";
					else
						return $"{NumberToText(Convert.ToInt32(sPesos)).Trim()}";
				}
			}
			catch (Exception ex)
			{
				class_Procedures.Show_Error(ex);
			}

			return null;
		}

		private static string NumberToText(int n)
		{
			switch (n)
			{
				case 0:
					{
						return string.Empty;
					}

				case var @case when 1 <= @case && @case <= 19:
					{
						var arr = new string[] { "One", "Two", "Three", "Four", "Five", "Six", "Seven", "Eight", "Nine", "Ten", "Eleven", "Twelve", "Thirteen", "Fourteen", "Fifteen", "Sixteen", "Seventeen", "Eighteen", "Nineteen" };
						return $"{arr[n - 1]}";
					}

				case var case1 when 20 <= case1 && case1 <= 99:
					{
						var arr = new string[] { "Twenty", "Thirty", "Forty", "Fifty", "Sixty", "Seventy", "Eighty", "Ninety" };
						return $"{arr[n / 10 - 2]} {NumberToText(n % 10)}";
					}

				case var case2 when 100 <= case2 && case2 <= 199:
					{
						return $"One Hundred {NumberToText(n % 100)}";
					}

				case var case3 when 200 <= case3 && case3 <= 999:
					{
						return $"{NumberToText(n / 100)} Hundred {NumberToText(n % 100)}";
					}

				case var case4 when 1000 <= case4 && case4 <= 1999:
					{
						return $"One Thousand {NumberToText(n % 1000)}";
					}

				case var case5 when 2000 <= case5 && case5 <= 999999:
					{
						return $"{NumberToText(n / 1000)} Thousand {NumberToText(n % 1000)}";
					}

				case var case6 when 1000000 <= case6 && case6 <= 1999999:
					{
						return $"One Million {NumberToText(n % 1000000)}";
					}

				case var case7 when 1000000 <= case7 && case7 <= 999999999:
					{
						return $"{NumberToText(n / 1000000)} Million {NumberToText(n % 1000000)}";
					}

				case var case8 when 1000000000 <= case8 && case8 <= 1999999999:
					{
						return $"One Billion {NumberToText(n % 1000000000)}";
					}

				default:
					{
						return $"{NumberToText(n / 1000000000)} Billion {NumberToText(n % 1000000000)}";
					}
			}
		}

		public static string GetOrdinal(int num)
		{
			switch (num % 100)
			{
				case 11:
				case 12:
				case 13:
					{
						return "th";
					}
			}

			switch (num % 10)
			{
				case 1:
					{
						return "st";
					}

				case 2:
					{
						return "nd";
					}

				case 3:
					{
						return "rd";
					}

				default:
					{
						return "th";
					}
			}
		}

		public static bool IsUpperCase(string input)
		{
			for (int i = 0, loopTo = input.Length - 1; i <= loopTo; i++)
			{
				if (string.Compare(input.Substring(i, 1), input.Substring(i, 1).ToUpper(), false) != 0)
					return false;
			}

			return true;
		}

		private static void Args_Showing(object sender, XtraMessageShowingArgs e)
		{
			// e.Form.Icon = XtraForm1.Icon
		}

		public static string Show_Input_Box(string sPrompt, string sCaption, string dDefaultResponse)
		{
			string sResult = string.Empty;
			var args = new XtraInputBoxArgs();
			args.Showing += Args_Showing;
			args.Caption = sCaption;
			args.Prompt = sPrompt;
			args.DefaultButtonIndex = 0;
			args.DefaultResponse = dDefaultResponse;
			try
			{
				sResult = XtraInputBox.Show(args).ToString();
			}
			catch (Exception)
			{
				return string.Empty;
			}

			class_Procedures.Copy_Clipboard(sResult);
			return sResult;
		}

		public static DateTime Show_Input_Box(string sPrompt, string sCaption, DateTime dDefaultResponse)
		{
			var args = new XtraInputBoxArgs();
			args.Showing += Args_Showing;
			args.Caption = sCaption;
			args.Prompt = sPrompt;
			args.DefaultButtonIndex = 0;
			var editor = new DateEdit();
			editor.Properties.CalendarView = DevExpress.XtraEditors.Repository.CalendarView.TouchUI;
			editor.Properties.Mask.EditMask = "MMMM dd, yyyy";
			args.Editor = editor;
			args.DefaultResponse = dDefaultResponse;
			return Convert.ToDateTime(XtraInputBox.Show(args));
		}

		public static int Show_Input_Box(string sPrompt, string sCaption, int dDefaultResponse)
		{
			var args = new XtraInputBoxArgs();
			args.Showing += Args_Showing;
			args.Caption = sCaption;
			args.Prompt = sPrompt;
			args.DefaultButtonIndex = 0;
			var editor = new CalcEdit();
			editor.Properties.Mask.EditMask = "N0";
			args.Editor = editor;
			args.DefaultResponse = dDefaultResponse;
			return Convert.ToInt32(XtraInputBox.Show(args));
		}

		public static decimal Show_Input_Box(string sPrompt, string sCaption, decimal dDefaultResponse)
		{
			var args = new XtraInputBoxArgs();
			args.Showing += Args_Showing;
			args.Caption = sCaption;
			args.Prompt = sPrompt;
			args.DefaultButtonIndex = 0;
			var editor = new CalcEdit();
			editor.Properties.Mask.EditMask = "N2";
			args.Editor = editor;
			args.DefaultResponse = dDefaultResponse;
			return Convert.ToDecimal(XtraInputBox.Show(args));
		}

		public static bool gridView_Loop(DevExpress.XtraGrid.Views.Grid.GridView gridView, int iRowHandleOld, int iRowHandle, string sTitle, string sAction, bool bLoopStop = false, bool bLoopLive = true, bool bSelectAll = false, string columnFieldName = "Notes")
		{
			if (bLoopStop || class_Database.DBError)
			{
				gridView.SetRowCellValue(iRowHandle, columnFieldName, class_Database.sError);
				string argsTitle = string.Empty;
				class_Procedures.Loop_Stop_Message(sTitle, sAction, sTitle: ref argsTitle);
				return false;
			}

			if (bLoopLive)
			{
				if (!bSelectAll)
					gridView.UnselectRow(iRowHandleOld);
				gridView.FocusedRowHandle = iRowHandle;
				gridView.MakeRowVisible(iRowHandle);
				if (!bSelectAll)
					gridView.SelectRow(iRowHandle);
				Application.DoEvents();
			}

			return true;
		}
	}
}

File: MyClassLibrary/Classes/class_Functions.vb
Imports DevExpress.XtraEditors
Imports MyClassLibrary.Classes
Imports System.Runtime.CompilerServices
Imports System.Text.RegularExpressions

Namespace Classes
    Public Module class_Functions

        Public Enum PasswordScore
            Blank
            VeryWeak
            Weak
            Medium
            Strong
            VeryStrong
        End Enum

        Public Function CheckPasswordStrength(ByVal password As String) As PasswordScore
            Dim N As Integer = 0
            Dim L As Integer = password.Length
            If L = 0 Then Return PasswordScore.Blank
            If Regex.IsMatch(password, "[\d]", RegexOptions.ECMAScript) Then N += 10
            If Regex.IsMatch(password, "[a-z]", RegexOptions.ECMAScript) Then N += 26
            If Regex.IsMatch(password, "[A-Z]", RegexOptions.ECMAScript) Then N += 26
            If Regex.IsMatch(password, "[~`!@#$%\^\&\*\(\)\-_\+=\[\{\]\}\|\\;:'\""<\,>\.\?\/£]", RegexOptions.ECMAScript) AndAlso password.Length > 8 Then N += 33
            Dim H As Integer
            Try
                H = Convert.ToInt32(L * (Math.Round(Math.Log(N) / Math.Log(2))))
            Catch Ex As OverflowException
                H = 0
                class_Procedures.Set_Message(class_Procedures.MsgMode.Sql, $"{Environment.NewLine}{Ex.Source}{Environment.NewLine}{Ex.Message}")
            End Try
            If H <= 32 Then Return PasswordScore.VeryWeak
            If H <= 48 Then Return PasswordScore.Weak
            If H <= 64 Then Return PasswordScore.Medium
            If H <= 80 Then Return PasswordScore.Strong
            Return PasswordScore.VeryStrong
        End Function

        Public Function Product_Name_Version() As String
            Return $"{Application.ProductName} {Application.ProductVersion}™ {Date.Today.Year}"
        End Function

        Public Function AutoID(ByVal sPreCode As String) As String
            Return $"{sPreCode}{DateTime.Now:yyyyMMddHHmmssfff)}"
        End Function

        Public Function FirstDayOfMonth(ByVal value As DateTime) As DateTime
            Return New DateTime(value.Year, value.Month, 1)
        End Function

        <Extension()>
        Public Function DaysInMonth(ByVal value As DateTime) As Integer
            Return DateTime.DaysInMonth(value.Year, value.Month)
        End Function

        <Extension()>
        Public Function LastDayOfMonth(ByVal value As DateTime) As DateTime
            Return New DateTime(value.Year, value.Month, value.DaysInMonth())
        End Function

        Public Function GetMonthDifference(ByVal startDate As Date, ByVal endDate As Date) As Integer
            Return Math.Abs(12 * (startDate.Year - endDate.Year) + startDate.Month - endDate.Month)
        End Function

        Public Function Next_Working_Date(ByVal dt As Date, Optional ByVal bShiftPressed As Boolean = False) As Date
            Dim i As Integer
            If bShiftPressed Then
                i = -1
            Else
                i = 1
            End If
            dt = dt.AddDays(i)
            If dt.DayOfWeek.Equals(DayOfWeek.Sunday) Then
                dt = dt.AddDays(i)
            End If
            Return dt
        End Function

        Public Function WorkingDays(ByVal dtDateFrom As Date, ByVal dtDateTo As Date) As Integer
            Dim _elapsedDays As Integer = 0
            Dim _weekendDays As DayOfWeek() = {DayOfWeek.Sunday}

            For i = 0 To (dtDateTo - dtDateFrom).Days
                If Not _weekendDays.Contains(dtDateFrom.AddDays(i).DayOfWeek) Then _elapsedDays += 1
            Next

            Return _elapsedDays
        End Function

        Public Function WorkingDays(ByVal dtDateFrom As Date, ByVal iDAys As Integer) As Date
            Dim _elapsedDate As Date = dtDateFrom
            Dim _weekendDays As DayOfWeek() = {DayOfWeek.Saturday, DayOfWeek.Sunday}

            If iDAys > 0 Then
                For i = 0 To iDAys - 1
                    _elapsedDate = _elapsedDate.AddDays(1)
                    If _weekendDays.Contains(_elapsedDate.DayOfWeek) Then
                        i -= 1
                    End If
                Next
            End If
            Return _elapsedDate
        End Function

        Public Function Generate_ID(ByVal sCode As String, ByVal iZeros As Integer) As String
            Return CInt(sCode.Remove(0, sCode.Length - iZeros)).ToString
        End Function

        Public Function Generate_ID(ByVal sPreCode As String, ByVal iZeros As Integer, ByVal iCode As Integer) As String
            Dim sTmpCode, sCode As String
            Dim iLength As Integer = iCode.ToString.Length

            If iLength > iZeros Then iZeros = iLength
            Dim sZeros As String = New String("0"c, iZeros)
            sTmpCode = $"{sPreCode}{sZeros}"
            sCode = sTmpCode.Remove(sTmpCode.Length - iLength)

            Return sCode.Insert(sTmpCode.Length - iLength, iCode.ToString)
        End Function

        Public Function Code_Increment(ByVal sCode As String, ByVal iCount As Integer) As String
            If sCode.Equals(String.Empty) Then Return String.Empty
            Dim iValueNew As Integer = CInt(sCode.Remove(0, sCode.Length - iCount)) + 1
            Return $"{sCode.Remove(sCode.Length - iValueNew.ToString.Length)}{iValueNew}"
        End Function

        Public Function calculate_VAT(ByVal dAmount As Decimal) As Decimal
            Return Math.Round((dAmount / CDec(1.12)) * CDec(0.12), 2)
        End Function

        Public Function calculate_no_VAT(ByVal dAmount As Decimal) As Decimal
            Return Math.Round(dAmount - ((dAmount / CDec(1.12)) * CDec(0.12)), 2)
        End Function

        Public Function Percent(ByVal dAmount As Decimal, ByVal dPercent As Decimal) As Decimal
            Return ((dAmount * dPercent) / 100) - dAmount
        End Function

        Public Function maskString(ByVal s As String, ByVal iUnMaskLeft As Integer, ByVal iUnMaskRight As Integer) As String
            Dim sReturn As String = Nothing
            Try
                Dim firstPart As String = s.Substring(0, iUnMaskLeft)
                Dim len As Integer = s.Length
                Dim lastPart As String = s.Substring(len - iUnMaskRight, iUnMaskRight)
                Dim middlePartLenght As Integer = len - (firstPart.Length + lastPart.Length)
                Dim middlePart As String = New String("•"c, middlePartLenght)
                sReturn = $"{firstPart}{middlePart}{lastPart}"
            Catch Ex As Exception
                class_Procedures.Show_Error(Ex)
            End Try
            Return sReturn
        End Function

        Public Function AmountToWords(ByVal sNumber As String) As String
            Dim sNumberSet As String = Trim(sNumber)
            Dim sAmount As String() = sNumberSet.Split(New Char() {"."c})
            Dim sPesos = sAmount(0)
            Dim sCents As String = Nothing

            Try
                sCents = Left(sAmount(1), 2)
            Catch Ex As Exception
                class_Procedures.Set_Message(class_Procedures.MsgMode.Sql, $"{Ex.Source}{Environment.NewLine}{Ex.Message}{Environment.NewLine}Please make sure that the [Amount in Words] are specified.", "Amount in Words")
            End Try

            Try
                If CInt(sCents) > 0 Then
                    Return $"{NumberToText(CInt(sPesos)).Trim} Pesos and {NumberToText(CInt(sCents))} Cents Only"
                Else
                    Return $"{NumberToText(CInt(sPesos)).Trim} Pesos Only"
                End If
            Catch Ex As OverflowException
                class_Procedures.Show_Error(Ex)
            End Try
            Return Nothing
        End Function

        Private Function NumberToText(ByVal n As Integer) As String
            Select Case n
                Case 0
                    Return String.Empty
                Case 1 To 19
                    Dim arr() As String = {"One", "Two", "Three", "Four", "Five", "Six", "Seven", "Eight", "Nine", "Ten", "Eleven", "Twelve", "Thirteen", "Fourteen", "Fifteen", "Sixteen", "Seventeen", "Eighteen", "Nineteen"}
                    Return $"{arr(n - 1)}"
                Case 20 To 99
                    Dim arr() As String = {"Twenty", "Thirty", "Forty", "Fifty", "Sixty", "Seventy", "Eighty", "Ninety"}
                    Return $"{arr(n \ 10 - 2)} {NumberToText(n Mod 10)}"
                Case 100 To 199
                    Return $"One Hundred {NumberToText(n Mod 100)}"
                Case 200 To 999
                    Return $"{NumberToText(n \ 100)} Hundred {NumberToText(n Mod 100)}"
                Case 1000 To 1999
                    Return $"One Thousand {NumberToText(n Mod 1000)}"
                Case 2000 To 999999
                    Return $"{NumberToText(n \ 1000)} Thousand {NumberToText(n Mod 1000)}"
                Case 1000000 To 1999999
                    Return $"One Million {NumberToText(n Mod 1000000)}"
                Case 1000000 To 999999999
                    Return $"{NumberToText(n \ 1000000)} Million {NumberToText(n Mod 1000000)}"
                Case 1000000000 To 1999999999
                    Return $"One Billion {NumberToText(n Mod 1000000000)}"
                Case Else
                    Return $"{NumberToText(n \ 1000000000)} Billion {NumberToText(n Mod 1000000000)}"
            End Select
        End Function

        Public Function IsUpperCase(ByVal input As String) As Boolean
            For i As Integer = 0 To input.Length - 1
                If String.Compare(input.Substring(i, 1), input.Substring(i, 1).ToUpper(), False) <> 0 Then Return False
            Next
            Return True
        End Function

        Private Sub Args_Showing(sender As Object, e As XtraMessageShowingArgs)
            'e.Form.Icon = XtraForm1.Icon
        End Sub

        Public Function Show_Input_Box(ByVal sPrompt As String, ByVal sCaption As String, ByVal dDefaultResponse As String) As String
            Dim sResult As String = String.Empty
            Dim args As New XtraInputBoxArgs()
            AddHandler args.Showing, AddressOf Args_Showing

            args.Caption = sCaption
            args.Prompt = sPrompt
            args.DefaultButtonIndex = 0
            args.DefaultResponse = dDefaultResponse

            Try
                sResult = XtraInputBox.Show(args).ToString
            Catch Ex As Exception
                Return String.Empty
            End Try

            class_Procedures.Copy_Clipboard(sResult)
            Return sResult
        End Function

        Public Function Show_Input_Box(ByVal sPrompt As String, ByVal sCaption As String, ByVal dDefaultResponse As Date) As Date
            Dim args As New XtraInputBoxArgs()
            AddHandler args.Showing, AddressOf Args_Showing

            args.Caption = sCaption
            args.Prompt = sPrompt
            args.DefaultButtonIndex = 0

            Dim editor As New DateEdit()
            editor.Properties.CalendarView = Repository.CalendarView.TouchUI
            editor.Properties.Mask.EditMask = "MMMM dd, yyyy"
            args.Editor = editor
            args.DefaultResponse = dDefaultResponse
            Return CDate(XtraInputBox.Show(args))
        End Function

        Public Function Show_Input_Box(ByVal sPrompt As String, ByVal sCaption As String, ByVal dDefaultResponse As Integer) As Integer
            Dim args As New XtraInputBoxArgs()
            AddHandler args.Showing, AddressOf Args_Showing

            args.Caption = sCaption
            args.Prompt = sPrompt
            args.DefaultButtonIndex = 0

            Dim editor As New CalcEdit()
            editor.Properties.Mask.EditMask = "N0"
            args.Editor = editor
            args.DefaultResponse = dDefaultResponse

            Return CInt(XtraInputBox.Show(args))
        End Function

        Public Function Show_Input_Box(ByVal sPrompt As String, ByVal sCaption As String, ByVal dDefaultResponse As Decimal) As Decimal
            Dim args As New XtraInputBoxArgs()
            AddHandler args.Showing, AddressOf Args_Showing

            args.Caption = sCaption
            args.Prompt = sPrompt
            args.DefaultButtonIndex = 0

            Dim editor As New CalcEdit()
            editor.Properties.Mask.EditMask = "N2"
            args.Editor = editor
            args.DefaultResponse = dDefaultResponse

            Return CDec(XtraInputBox.Show(args))
        End Function

        Public Function GridView_Loop(ByVal gridView As DevExpress.XtraGrid.Views.Grid.GridView, ByVal iRowHandleOld As Integer, ByVal iRowHandle As Integer, ByVal sTitle As String, ByVal sAction As String, ByVal bLoopStop As Boolean, Optional ByVal bLoopLive As Boolean = True, Optional ByVal bSelectAll As Boolean = False, Optional ByVal columnFieldName As String = "Notes") As Boolean
            With gridView
                If bLoopStop OrElse class_Database.DBError Then
                    .SetRowCellValue(iRowHandle, columnFieldName, class_Database.sError)
                    class_Procedures.Loop_Stop_Message(sTitle, sAction)
                    Return False
                End If

                If bLoopLive Then
                    If Not bSelectAll Then .UnselectRow(iRowHandleOld)
                    .FocusedRowHandle = iRowHandle
                    .MakeRowVisible(iRowHandle)
                    If Not bSelectAll Then .SelectRow(iRowHandle)
                    Application.DoEvents()
                End If
            End With
            Return True
        End Function

    End Module
End Namespace